In [13]:
import cv2,os
import numpy as np


data_path=r'D:\Data_Science\DL_Workshop\Rewatch\Day 05\PetImages'


categories=os.listdir(data_path)
labels=np.arange(len(categories))

label_dict=dict(zip(categories,labels))

print(label_dict)
print(categories)
print(labels)

{'Cat': 0, 'Dog': 1}
['Cat', 'Dog']
[0 1]


In [14]:
img_size=50

data=[]
target=[]

for category in categories:
    folder_path=os.path.join(data_path,category)
    img_names=os.listdir(folder_path)
        
    for img_name in img_names:
        img_path=os.path.join(folder_path,img_name)
         
        try:
            img=cv2.imread(img_path)
            #cv2.imshow('LIVE',img)
            #cv2.waitKey(100)
            
            gray=cv2.cvtColor(img,cv2.COLOR_BGR2GRAY)
            #Coverting the image into gray scale
            resized=cv2.resize(gray,(img_size,img_size))
            #resizing the gray scale into 50x50, since we need a fixed common size for all the images in the dataset
            data.append(resized)
            target.append(label_dict[category])
            
        except Exception as e:
            print(img_path)
            print(e)
            #if any exception rasied, the exception will be printed here. And pass to the next image

D:\Data_Science\DL_Workshop\Rewatch\Day 05\PetImages\Cat\10125.jpg
OpenCV(4.5.1) C:\Users\appveyor\AppData\Local\Temp\1\pip-req-build-kh7iq4w7\opencv\modules\imgproc\src\color.cpp:182: error: (-215:Assertion failed) !_src.empty() in function 'cv::cvtColor'

D:\Data_Science\DL_Workshop\Rewatch\Day 05\PetImages\Cat\10404.jpg
OpenCV(4.5.1) C:\Users\appveyor\AppData\Local\Temp\1\pip-req-build-kh7iq4w7\opencv\modules\imgproc\src\color.cpp:182: error: (-215:Assertion failed) !_src.empty() in function 'cv::cvtColor'

D:\Data_Science\DL_Workshop\Rewatch\Day 05\PetImages\Cat\10501.jpg
OpenCV(4.5.1) C:\Users\appveyor\AppData\Local\Temp\1\pip-req-build-kh7iq4w7\opencv\modules\imgproc\src\color.cpp:182: error: (-215:Assertion failed) !_src.empty() in function 'cv::cvtColor'

D:\Data_Science\DL_Workshop\Rewatch\Day 05\PetImages\Cat\10820.jpg
OpenCV(4.5.1) C:\Users\appveyor\AppData\Local\Temp\1\pip-req-build-kh7iq4w7\opencv\modules\imgproc\src\color.cpp:182: error: (-215:Assertion failed) !_src.empt

D:\Data_Science\DL_Workshop\Rewatch\Day 05\PetImages\Dog\11849.jpg
OpenCV(4.5.1) C:\Users\appveyor\AppData\Local\Temp\1\pip-req-build-kh7iq4w7\opencv\modules\imgproc\src\color.cpp:182: error: (-215:Assertion failed) !_src.empty() in function 'cv::cvtColor'

D:\Data_Science\DL_Workshop\Rewatch\Day 05\PetImages\Dog\11853.jpg
OpenCV(4.5.1) C:\Users\appveyor\AppData\Local\Temp\1\pip-req-build-kh7iq4w7\opencv\modules\imgproc\src\color.cpp:182: error: (-215:Assertion failed) !_src.empty() in function 'cv::cvtColor'

D:\Data_Science\DL_Workshop\Rewatch\Day 05\PetImages\Dog\1308.jpg
OpenCV(4.5.1) C:\Users\appveyor\AppData\Local\Temp\1\pip-req-build-kh7iq4w7\opencv\modules\imgproc\src\color.cpp:182: error: (-215:Assertion failed) !_src.empty() in function 'cv::cvtColor'

D:\Data_Science\DL_Workshop\Rewatch\Day 05\PetImages\Dog\1866.jpg
OpenCV(4.5.1) C:\Users\appveyor\AppData\Local\Temp\1\pip-req-build-kh7iq4w7\opencv\modules\imgproc\src\color.cpp:182: error: (-215:Assertion failed) !_src.empty(

### Exception Handling

In the previous code segment, a try and except blocks are included in order to handle any raised exception while reading the images. These exceptions may raised due to few possible reasons

1. Files which are not a images (jpg,jpeg,png etc.)
2. Courrupted files

In [15]:
import numpy as np

data=np.array(data)
target=np.array(target)

In [16]:
print(data.shape,target.shape)

(24946, 50, 50) (24946,)


Checking the size of the dataset, without few corrupted images the size shoulde be around 24950, the original dataset size was 25000.

Note that the dataset has 24946 rows and 2 columns. The first and second columns contain the resized image and the categorical label respectively. Above code segment split the dataset into data(images) and target(categorical labels)

In [17]:
data_new=data/255.0
data_new=data_new.reshape(data_new.shape[0],img_size,img_size,1)

from keras.utils import np_utils

target_new=np_utils.to_categorical(target)

### data and target preprocessing

The above code segment perform some necessary proprocessing task for data and target

1.```data=np.array(data)/255.0``` - Dividing all the pixels in all the images by 255 in order to convert them to the range 0-1.
This is to reduce the complexity in training the Neural Network.

2.```data=np.reshape(data,(data.shape[0],img_size,img_size,1))``` - The CNNs require 4 dimensional input, therefore we add the  dimension 1, this is to represent grayscale image, for color images you may have to use 3.

3.```new_target=np_utils.to_categorical(target)``` - Converting the labels into categorical representation.

In [18]:
np.save('data',data_new)
np.save('target',target_new)

saving data and target as numpy array files

In [20]:
target_new.shape

(24946, 2)